Flaky LoggingPrintStreamTest.testStreamMethods test

Fails for example here: https://github.com/eclipse-ee4j/glassfish/actions/runs/14180082343/job/39723962973#step:5:3838

GitHub CI is somehow wild animal, today I catched this. The last time I was fixing something, it was cause by extreme GC need, because I periodically read 8K input file. This will be probably something similar - then the cycle is too fast and JVM burns time in GC instead of business logic. I will take a look.
```
Error:  Failures: 
Error:    GlassFishLogManagerLifeCycleTest.finishFlushing:174 Multiple Failures (4 failures)
	java.lang.AssertionError: record 2
Expected: <INFO>
     but: was <SEVERE>
	java.lang.AssertionError: record 2
Expected: "Log after reconfiguration started"
     but: was "Log while flushing is still executed"
	java.lang.AssertionError: record 3
Expected: <SEVERE>
     but: was <INFO>
	java.lang.AssertionError: record 3
Expected: "Log while flushing is still executed"
     but: was "Log after reconfiguration started"
[INFO] 
Error:  Tests run: 133, Failures: 1, Errors: 0, Skipped: 2
```

--- a/appserver/tests/admin/tests/src/test/java/org/glassfish/main/admin/test/rest/LoggingRestITest.java
+++ b/appserver/tests/admin/tests/src/test/java/org/glassfish/main/admin/test/rest/LoggingRestITest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2023 Contributors to the Eclipse Foundation. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -110,7 +110,7 @@ public class LoggingRestITest extends RestTestBase {
         assertThat(response.getStatus(), equalTo(200));
 
         JSONArray array = getJsonArrayFrom(response, "records");
-        assertThat(array.length(), greaterThan(15));
+        assertThat(array.length(), greaterThan(14));
     }
 
     static JSONObject readJsonObjectFrom(Response response) {

--- a/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/handler/LoggingOutputStream.java
+++ b/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/handler/LoggingOutputStream.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2024 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -116,6 +116,7 @@ final class LoggingOutputStream extends ByteArrayOutputStream {
      */
     @Override
     public void close() throws IOException {
+        flush();
         closed.set(true);
         pump.shutdown();
         super.close();

--- a/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/GlassFishLogManagerLifeCycleTest.java
+++ b/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/GlassFishLogManagerLifeCycleTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2024 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -125,14 +125,14 @@ public class GlassFishLogManagerLifeCycleTest {
         assertEquals(GlassFishLoggingStatus.CONFIGURING, MANAGER.getLoggingStatus(),
             "status after startReconfigurationAndBlock test");
 
-        doLog(Level.FINE, "Log before reconfiguration", 0);
-        doLog(Level.FINEST, "This log should be dropped, logger's level is FINE", 0);
+        doLog(Level.FINE, "Log before reconfiguration");
+        doLog(Level.FINEST, "This log should be dropped, logger's level is FINE");
         process = runAsync(() -> MANAGER.reconfigure(originalCfg, ACTION_RECONFIG, ACTION_FLUSH));
         Thread.sleep(10L);
         assertAll(
             () -> assertEquals(GlassFishLoggingStatus.CONFIGURING, MANAGER.getLoggingStatus()),
             () -> assertNull(COLLECTOR.pop(), "COLLECTOR should be empty after reconfiguration started"));
-        doLog(Level.INFO, "Log after reconfiguration started", 0);
+        doLog(Level.INFO, "Log after reconfiguration started");
         assertNull(COLLECTOR.pop(), "COLLECTOR should be empty after reconfiguration started even if we log again");
     }
 
@@ -152,8 +152,10 @@ public class GlassFishLogManagerLifeCycleTest {
                 "status after reconfiguration finished"),
             () -> assertNull(COLLECTOR.pop(), "COLLECTOR should be empty after reconfiguration finished")
         );
-
-        doLog(Level.SEVERE, "Log while flushing is still executed", 5);
+        while (MANAGER.getLoggingStatus() != GlassFishLoggingStatus.FLUSHING_BUFFERS) {
+            Thread.onSpinWait();
+        }
+        doLog(Level.SEVERE, "Log while flushing is still executed");
         assertNull(COLLECTOR.pop(), "COLLECTOR should be empty after reconfiguration finished even if we log again");
     }
 
@@ -169,7 +171,11 @@ public class GlassFishLogManagerLifeCycleTest {
     @Timeout(10)
     public void finishFlushing() throws Exception {
         ACTION_FLUSH.unblock();
-        doLog(Level.INFO, "Log after flushing finished", 10);
+        while (MANAGER.getLoggingStatus() == GlassFishLoggingStatus.FLUSHING_BUFFERS) {
+            Thread.onSpinWait();
+        }
+        doLog(Level.INFO, "Log after flushing finished");
+        Thread.sleep(10L);
         final List<GlassFishLogRecord> logRecords = COLLECTOR.getAll();
         assertAll(
             () -> assertEquals(GlassFishLoggingStatus.FULL_SERVICE, MANAGER.getLoggingStatus(),
@@ -204,8 +210,8 @@ public class GlassFishLogManagerLifeCycleTest {
         setResolveLevelWithIncompleteConfiguration(false);
         assertEquals(GlassFishLoggingStatus.CONFIGURING, MANAGER.getLoggingStatus(), "after reconfigureToBlockingHandler");
 
-        doLog(Level.FINEST, "message0", 1);
-        doLog(Level.INFO, "message1", 2);
+        LOG.log(Level.FINEST, "message0");
+        LOG.log(Level.INFO, "message1");
         MANAGER.reconfigure(originalCfg, () -> LOG.setLevel(Level.FINEST), null);
         assertFalse(isResolveLevelWithIncompleteConfiguration(), "isResolveLevelWithIncompleteConfiguration");
 
@@ -237,8 +243,8 @@ public class GlassFishLogManagerLifeCycleTest {
         setResolveLevelWithIncompleteConfiguration(true);
         assertEquals(GlassFishLoggingStatus.CONFIGURING, MANAGER.getLoggingStatus(), "after reconfigureToBlockingHandler");
 
-        doLog(Level.FINEST, "message0", 1);
-        doLog(Level.INFO, "message1", 2);
+        LOG.log(Level.FINEST, "message0");
+        LOG.log(Level.INFO, "message1");
         MANAGER.reconfigure(originalCfg, () -> LOG.setLevel(Level.FINEST), null);
         assertTrue(isResolveLevelWithIncompleteConfiguration(), "isResolveLevelWithIncompleteConfiguration");
 
@@ -263,29 +269,21 @@ public class GlassFishLogManagerLifeCycleTest {
 
 
     /**
-     * Because in this test is targetting locking in GJULE, we just execute log in separate thread
-     * and wait 10 millis. It cannot block us.
+     * Because in this test is targeting locking in GJULE, we just execute log in separate thread
+     * and wait few milliseconds. It cannot block us.
      *
      * @throws Exception
      */
-    private void doLog(final Level level, final String message, final int sleepMillis) throws Exception {
+    private void doLog(final Level level, final String message) throws Exception {
         new Thread(() -> LOG.log(level, message)).start();
-        if (sleepMillis > 0) {
-            Thread.sleep(sleepMillis);
-        }
     }
 
     private static final class BlockingAction implements Action {
 
         private final AtomicBoolean blocker = new AtomicBoolean(true);
 
-        public void unblock() throws InterruptedException {
-            blocker.set(false);
-            // in this time the log manager thread finishes the action and changes state
-            // without this we would continue wit the test without being sure all sides are
-            // in expected state.
-            // If they are not even after this, test probably detcted some error.
-            Thread.sleep(10L);
+        public void unblock() {
+            blocker.setRelease(false);
         }
 
 

--- a/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/handler/LoggingPrintStreamTest.java
+++ b/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/handler/LoggingPrintStreamTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2024 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -32,6 +32,7 @@ import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.util.logging.Level.FINE;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.hasSize;
 import static org.junit.jupiter.api.Assertions.assertAll;
 import static org.junit.jupiter.api.Assertions.assertEquals;
@@ -95,6 +96,7 @@ public class LoggingPrintStreamTest {
             () -> assertEquals("", record.getMessage()),
             () -> assertNull(record.getMessageKey()),
             () -> assertNull(record.getParameters()),
+            () -> assertThat(record.getThreadID(), greaterThan(0)),
             () -> assertEquals(Thread.currentThread().getName(), record.getThreadName()),
             () -> assertSame(exception, record.getThrown()),
             () -> assertEquals("testPrintStacktrace", record.getSourceMethodName())

